db.go

```diff
@@ -53,14 +54,10 @@ type DBResponse struct {
 }
 
 func newDBRunner(name, dsn string) (*dbRunner, error) {
-	nx, err := connectDB(dsn)
```

Comment:
Because newDBRunner() is executed for each runbook, how about generating a random DSN here for spanner only?

Reply:
I think we don't have to worry about runbook IDs either.

Reply:
You are right. Using randomly generated numbers should work fine and I originally used them for this purpose. However, I changed my implementation because I thought using random numbers is not robust and it would be better to avoid if possible.

How do you think about the following code which is my initial workaround implementation.
```go
db, err = sql.Open("spanner", fmt.Sprintf(`projects/%s/instances/%s/databases/%s;workaroundConnectionInvalidationKey=%d`, d[0], d[1], d[2], rand.Int63()))
```

Reply:
> because I thought using random numbers is not robust and it would be better to avoid if possible.

Indeed, I understand.

So how about [xid](https://github.com/rs/xid), which used to be used to generate IDs for runn?

https://github.com/k1LoW/runn/blob/4df90a92a72f4d684d86b79ad6f15773fe9dd1da/operator.go#L1161-L1163

It is more robust than random numbers because it uses time.

Reply:
Thanks, it seems xid is a good alternative! It has a **3-byte counter** part instead of the random number.

> 4-byte value representing the **seconds** since the Unix epoch,
> 3-byte machine identifier,
> 2-byte process id, and
> **3-byte counter**, starting with a random value

---

> It is more robust than random numbers because it uses time.

I think xid does offer better duplication resistance in theory, but not because using the time in this situation. The xid uses **second precision** unix epoch, however the `newDBRunner()` function is invoked for every runbooks at once during setup phase. It seems very large number of `newDBRunner()` function can be invoked within a second in contrast to the `func (rnr *dbRunner) run(ctx context.Context, q *dbQuery, s *step) error` function. The 3-byte counter part is more important in this situation.

Reply:
Oops, second precision is not good.

I guess it would be better to use UUID or https://github.com/oklog/ulid or simply nano-precision time.

Reply:
I think xid is **good enough** in this situation. It has 3-bytes counter part, so it can be duplicated only if more than `2**24 = 16777216` xids generated per second.

```
1[sec] / (2^24)  =  1 [sec] / 16777216  ≒  60 [ns]
```

---

> use UUID or https://github.com/oklog/ulid 

I do not want to add more dependencies only for this workaround purpose...

---

> simply nano-precision time

I think this is difficult in practice. The timestamp resolution depends on the hardware and OS and [it does not have nano-precision time](https://stackoverflow.com/questions/57285292/why-does-time-now-unixnano-returns-the-same-result-after-an-io-operation) in most of time.

If you think 3 bytes counter is not enough, I recommend you to use a [64 bit atomic counter](https://pkg.go.dev/sync/atomic#Uint64.Add) instead of depending on timestamps.

Reply:
> I do not want to add more dependencies only for this workaround purpose...

> I think this is difficult in practice. The timestamp resolution depends on the hardware and OS and [it does not have nano-precision time](https://stackoverflow.com/questions/57285292/why-does-time-now-unixnano-returns-the-same-result-after-an-io-operation) in most of time.

Thank you!
You are absolutely right. I am learning a lot.

> I recommend you to use a [64 bit atomic counter](https://pkg.go.dev/sync/atomic#Uint64.Add) instead of depending on timestamps.

GREAT!  I agree with this idea because it is simple.

Reply:
@k1LoW I have updated the code to use 64 bit atomic counter 😉